contracts/modules/royalty/policies/IpRoyaltyVault.sol

```diff
@@ -394,4 +399,78 @@ contract IpRoyaltyVault is IIpRoyaltyVault, ERC20SnapshotUpgradeable, Reentrancy
             $.slot := IpRoyaltyVaultStorageLocation
         }
     }
+
+    function claimRevenueOnBehalfByTokenBatch(
+        address[] calldata tokenList,
+        address claimer
+    ) external nonReentrant whenNotPaused returns (uint256[] memory) {
+        IpRoyaltyVaultStorage storage $ = _getIpRoyaltyVaultStorage();
+
+        if (ROYALTY_MODULE.isIpRoyaltyVault(claimer) && msg.sender != claimer)
+            revert Errors.IpRoyaltyVault__VaultsMustClaimAsSelf();
+
+        if (IP_ASSET_REGISTRY.isWhitelistedGroupRewardPool(claimer) && msg.sender != GROUPING_MODULE)
+            revert Errors.IpRoyaltyVault__GroupPoolMustClaimViaGroupingModule();
+
+        uint256[] memory claimableAmounts = new uint256[](tokenList.length);
+        for (uint256 i = 0; i < tokenList.length; i++) {
+            claimableAmounts[i] = _clearPendingRewards(claimer, tokenList[i]);
+            if (claimableAmounts[i] == 0) revert Errors.IpRoyaltyVault__NoClaimableTokens();
+            $.userRewardInfo[tokenList[i]][claimer] += claimableAmounts[i];
+
+            $.claimVaultAmount[tokenList[i]] -= claimableAmounts[i];
+
+            emit RevenueTokenClaimed(claimer, tokenList[i], claimableAmounts[i]);
+        }
+
+        return claimableAmounts;
+    }
+
+    function pendingRewards(address user, address token) external view returns (uint256) {
+        return _pendingRewards(user, token);
+    }
+
+    function _transfer(address from, address to, uint256 amount) internal override {
+        // when transferring RoyaltyTokens (Vault) from a user to another user:
+        // 1. clear pending rewards of the (from) user
+        // 2. clear pending rewards of the (to) another user, if pending rewards of another user is not zero
+        // 3. update the rewardDebt of the (to) another user to accBalancePerShare * userAmount
+        // 4. update the rewardDebt of the (from) user to accBalancePerShare * userAmount
+        // 5. transfer the RoyaltyTokens (Vault) from the (from) user to the (to) another user
+        if (balanceOf(from) == 0) revert Errors.IpRoyaltyVault__ZeroBalance(address(this), from);
+        if (balanceOf(from) < amount) revert Errors.IpRoyaltyVault__InsufficientBalance(address(this), from, amount);
+
+        IpRoyaltyVaultStorage storage $ = _getIpRoyaltyVaultStorage();
+        address[] memory tokenList = $.tokens.values();
+        for (uint256 i = 0; i < tokenList.length; i++) {
+            address token = tokenList[i];
+            _clearPendingRewards(from, token);
+            _clearPendingRewards(to, token);
+            $.userRewardInfo[token][to] = ($.poolInfo[token] * (balanceOf(to) + amount)) / totalSupply();
+            $.userRewardInfo[token][from] = ($.poolInfo[token] * (balanceOf(from) - amount)) / totalSupply();
```

Comment:
save gas by caching totalSupply in memory to reduce storage calls

Reply:
Yes, good catch. totalSupply() should be cached in memory.
The PR was a draft for quick demonstrating the new simplified logic, not yet polished.
The totalSupply has been cached in the latest commit.